Allow Freezing of Module containing interface attribute

bzinodev opened the pull request:
This patch allows to freeze model that utilizes interfaces. Freezing works
under the user assumption that the interfase module dones not aliases with
any value used in the model.

To enable freezing of such modules, added an extra pramater:

torch._C._freeze_module(module, ignoreInterfaces = True)

eellison commented on test/jit/test_module_interface.py:
```diff
@@ -453,6 +453,99 @@ def forward(self, x):
+        m.proxy_mod = m.sub
+        m.sub.b = m.proxy_mod.b
+        m.eval()
+        #with self.assertRaisesRegex(RuntimeError, "Interface module contains mutating attribute"):
```
Accidentally still here ?

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -291,6 +306,91 @@ class AttributePropagator {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr) {
+          return false;
```
Here were not checking if it contains an interface call

eellison reviewed (commented):
I’m not sure if we’re necessarily doing it “safe” currently - for instance, we don’t check if the top level module has any setAttrs that reassign the interface, and we don’t check that any of the interface calls contain nester setAttrs.

Having said that, interfaces aren’t a public api and I don’t know if it’s worth it to spend too much time on it. 

I think as a simple thing we could just inline all of interfaces first, and then run freezing as normal


**[Zino:]** Changed the patch to first inline interface calls then perform freezing. I have made changes so we don't check for setattr inside interfaces. Instead fail if freezing is unable to fold away the interface

bzinodev commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -507,9 +507,12 @@ void AliasDb::analyzeImpl(Node* node) {
-        return analyzeCreator(node);
+      if (isFrozen_ && node->kind() == prim::GetAttr) {
+        auto& ty = node->input()->type();
+        if (ty->cast<InterfaceType>() || ty->expect<ClassType>()->is_module()) {
```
We do need it bc of these "%1 : <InterfaceType> = getAttr("..."). Inlining does not make them disappear

eellison replied:
I'm not sure I understand still. When we inline interfaces we should be replacing the interface type with the module type

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -291,6 +306,91 @@ class AttributePropagator {
 
+  // In practice interface module should contain no setattr.
+  // This checks for setattr so freezing is failed.
+  bool safeInterfaceToFreeze(Function& function) {
```
Correct! but I think it is good error message. If I relax then I have to change aliasanalysis. I did not want to do it 

RuntimeError: self->type()->kind() == TypeKind::ClassType INTERNAL ASSERT FAILED at "../torch/csrc/jit/ir/alias_analysis.cpp":847, please report a bug to PyTorch.

I could change the assertion to 
(!isFrozen_ && self->type()->kind() == TypeKind::ClassType).what do you think?

bzinodev replied:
Changed the assertion to: 
self->type()->kind() == TypeKind::ClassType || self->type()->kind() == TypeKind::InterfaceType

eellison replied:
Personally I would rather provide no safety than half-safety re interfaces

eellison commented on test/jit/test_module_interface.py:
```diff
@@ -453,6 +453,99 @@ def forward(self, x):
+            def getb(self, x):
+                return self.b
+
+        class OrigMod(torch.nn.Module):
```
```
class Mod(torch.nn.Module):
     def __init__(self):
          self.a : Interface = a
          self.b : Interface = b

     def forward(self, input):
          self.a = b
          return self.b(input)
```

bzinodev replied:
I added similar test.

bzinodev committed:
Allow Freezing of Module containing interface attribute

  This patch allows to freeze model that utilizes interfaces. Freezing works
  under the user assumption that the interfase module dones not aliases with
  any value used in the model.

  To enable freezing of such modules, added an extra pramater:

    torch._C._freeze_module(module, freezeInterfaces = True)

  By default freezeInterfaces is to True.

eellison commented on test/jit/test_module_interface.py:
```diff
@@ -451,8 +451,150 @@ def forward(self, x):
+                self.b = torch.tensor([1.5])
+
+            def forward(self, x):
+                self.b[0] += 2;
```
This is the test case I was talking about, but it's only failing bc `self.b[0] += 2;`, which is an unrelated issue

bzinodev replied:
I deleted that line and it still failed. I will upload an update. 

tdlr;  we check attr.type() for mutability. if it is mutable and if moduel.type()->getattribute(name)  then we fail.   In your example self.a has to be preserved, and it is an interface type then it is failed.

eellison commented on test/jit/test_module_interface.py:
```diff
@@ -451,8 +451,150 @@ def forward(self, x):
+        out_f = mf.forward(input)
+        self.assertEqual(out_s, out_f)
+
+    def test_freeze_module_with_interface2(self):
```
nit: could you edit the test name so it's more clear what is being tested ?

eellison reviewed (commented):
```
class Mod(torch.nn.Module):
     def __init__(self):
          self.a : Interface = a
          self.b : Interface = b

     def forward(self, input):
          self.a = b
          return self.a(input)
```

Presently we would inline the forward of `a` and then invoke it with `b` module. I think for safety it would be good to check that there are no setAttrs of an interface type in the forward function of the top level module or interfaces we've inlined. 

[Zino] I add a test to check such model. Currently  freezing fails because self.a cannot be folded away.  Attribute 'a' has to be preserved  and we dont allow to preserve an interface attribute (for the moment).

eellison reviewed (approved):
LGTM. This may not be completely safe re: interfaces but it provides a good amount of safety. interfaces are not a public feature so I think this is fine; doing too much work on this is not worth our time atm. If we decide to make interfaces public we should revisit this.
